src/main/java/ch/njol/skript/util/ItemSource.java

```diff
@@ -0,0 +1,47 @@
+package ch.njol.skript.util;
+
+import ch.njol.skript.bukkitutil.ItemUtils;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+import ch.njol.skript.util.slot.Slot;
+import ch.njol.skript.aliases.ItemType;
+
+/**
+ * Container class for containing the origin of a {@link Slot}, {@link ItemStack}, and {@link ItemType}.
+ */
+public class ItemSource<T> {
+
+	private final T source;
+	private final ItemStack itemStack;
+
+	public ItemSource(T source) {
+		ItemStack itemStack = ItemUtils.asItemStack(source);
```

Comment:
I'm worried this also can lead to disconnection between the actual source object and whatever this returns. Generally wrappers want to limit the amount of state they contain and reference the wrapped object as much as possible.